flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/ExactlyOnceKafkaWriter.java

```diff
@@ -57,6 +58,7 @@
 
 import static org.apache.flink.util.IOUtils.closeAll;
 import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
 
 /**
```

Comment:
Can you create a high-level doc string about how the transaction mechanism works that future readers do not have to fully reverse engineer the code? :)

Reply:
I added a new class TransactionOwnership that is used to document things but also to check the migration path.